File: 42-trapping-rain-water/42-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        Deque<Integer> stack = new ArrayDeque<>();
        int res =0;
        for(int i = 0; i < height.length; i++) {
            while(!stack.isEmpty() && height[i] > height[stack.peekLast()]) {
                int lowIndex = stack.pollLast();
                if(stack.isEmpty()) break;
                int wall = Math.min(height[i], height[stack.peekLast()]);
                int width = i - stack.peekLast() - 1;
                res +=(wall - height[lowIndex]) * width;
            }
            stack.offerLast(i);
        }
        return res;
    }
}

File: 42-trapping-rain-water/42-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        l, r = 0, len(height) - 1
        lmax, rmax = height[l], height[r]
        res = 0
        while(l < r):
            if height[l] <= height[r]:
                if height[l] < lmax:
                    res += lmax - height[l]
                else:
                    lmax = height[l]
                l += 1
            else:
                if height[r] < rmax:
                    res += rmax - height[r]
                else:
                    rmax = height[r]
                r -= 1
        return res
                    